Channelz4 - channelz metric data collection

lyuxuan opened the pull request:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/lyuxuan/grpc-go/1)
<!-- Reviewable:end -->

lyuxuan committed:
init

lyuxuan committed:
init

lyuxuan committed:
add entity registration methods and tests.

lyuxuan committed:
add entity registration methods and tests.

lyuxuan committed:
add entity registration methods and tests.

lyuxuan committed:
add entity registration methods and tests.

lyuxuan committed:
propagate nested channel info through context. added test file.

lyuxuan committed:
propagate nested channel info through context. added test file.

lyuxuan committed:
propagate nested channel info through context. added test file.

lyuxuan committed:
propagate nested channel info through context. added test file.

lyuxuan committed:
copyright

lyuxuan committed:
copyright

lyuxuan committed:
move ChannelzOn to chanelz package. add leak check to test.

lyuxuan committed:
move ChannelzOn to chanelz package. add leak check to test.

lyuxuan committed:
move ChannelzOn to chanelz package. add leak check to test.

lyuxuan committed:
move ChannelzOn to chanelz package. add leak check to test.

lyuxuan committed:
client side delete

lyuxuan committed:
client side delete

lyuxuan committed:
avoid calling channelz functions inside entity's lock

lyuxuan committed:
avoid calling channelz functions inside entity's lock

lyuxuan committed:
avoid calling channelz functions inside entity's lock

lyuxuan committed:
avoid calling channelz functions inside entity's lock

lyuxuan committed:
temp

lyuxuan committed:
temp

lyuxuan committed:
avoid race

lyuxuan committed:
avoid race

lyuxuan committed:
avoid race

lyuxuan committed:
avoid race

lyuxuan committed:
rebase fix

lyuxuan committed:
rebase fix

lyuxuan committed:
add synchronization mechanism to avoid false postive race detection

lyuxuan committed:
add synchronization mechanism to avoid false postive race detection

lyuxuan committed:
add deletion test

lyuxuan committed:
add deletion test

lyuxuan committed:
resolve race (put registration inside entity's lock to prevent race on id w/r)

lyuxuan committed:
resolve race (put registration inside entity's lock to prevent race on id w/r)

lyuxuan committed:
merge RegisterXXX() with AddChild(). Directly export GetXXX().

lyuxuan committed:
merge RegisterXXX() with AddChild(). Directly export GetXXX().

lyuxuan committed:
benchmark

lyuxuan committed:
benchmark

lyuxuan committed:
relocate id declaration

lyuxuan committed:
relocate id declaration

lyuxuan committed:
fix review. update entry recursive deletion logic. improve function naming.

lyuxuan committed:
fix review. update entry recursive deletion logic. improve function naming.

lyuxuan committed:
fix review comments

lyuxuan committed:
fix review comments

lyuxuan committed:
fix review (add DialOption for channelz parent ID)

lyuxuan committed:
fix review (add DialOption for channelz parent ID)

lyuxuan committed:
add SetChannelzID() func to prevent race

lyuxuan committed:
pick up from channelz3

menghanl committed:
examples: Replace context.Background with context.WithTimeout (#1877)

lyuxuan committed:
Merge branch 'channelz2' into channelz4

lyuxuan committed:
fix minor bug

lyuxuan committed:
Merge branch 'channelz2' into channelz4

lyuxuan committed:
fix race for acw.ac access

menghanl commented:
Reviewed 9 of 17 files at r1.
Review status: 9 of 17 files reviewed at latest revision, 4 unresolved discussions.

---

*[clientconn.go, line 872 at r1](https://reviewable.io:443/reviews/lyuxuan/grpc-go/1#-L6U1sLG0yfTWEBMTokj:-L6U1sLG0yfTWEBMTokk:bx05o82) ([raw file](https://github.com/lyuxuan/grpc-go/blob/6a52ab13a3232d9f8ed5b670459ef2684b318aee/clientconn.go#L872)):*
> ```Go
> 
> // SetChannelzID sets the channelzID field.
> // This function should not be used by grpc user.
> ```

Why is this function needed? It looks a bit wired to me even for tests...

And this shouldn't be part of the channelz public interface.

---

*[picker_wrapper.go, line 67 at r1](https://reviewable.io:443/reviews/lyuxuan/grpc-go/1#-L6U3y9QOM-njGIkXGZ-:-L6U3y9QOM-njGIkXGZ0:b-wyqk5j) ([raw file](https://github.com/lyuxuan/grpc-go/blob/6a52ab13a3232d9f8ed5b670459ef2684b318aee/picker_wrapper.go#L67)):*
> ```Go
> 	ac.incrCallsStarted()
> 	return func(b balancer.DoneInfo) {
> 		if b.Err != nil && b.Err != io.EOF {
> ```

The `io.EOF` check should not be necessary any more.

---

*[stream.go, line 119 at r1](https://reviewable.io:443/reviews/lyuxuan/grpc-go/1#-L6U6uzCXunkmvNMM80V:-L6U6uzCXunkmvNMM80W:b-r3dhee) ([raw file](https://github.com/lyuxuan/grpc-go/blob/6a52ab13a3232d9f8ed5b670459ef2684b318aee/stream.go#L119)):*
> ```Go
> func newClientStream(ctx context.Context, desc *StreamDesc, cc *ClientConn, method string, opts ...CallOption) (_ ClientStream, err error) {
> 	if channelz.IsOn() {
> 		cc.incrCallsStarted()
> ```

Is it possible to move cc.incrStart/cc.incrEnd into ac.incrStart/ac.incrEnd?

---

*[transport/control.go, line 336 at r1](https://reviewable.io:443/reviews/lyuxuan/grpc-go/1#-L6U83kI6gEy7XgGYt67:-L6U83kI6gEy7XgGYt68:bufucq) ([raw file](https://github.com/lyuxuan/grpc-go/blob/6a52ab13a3232d9f8ed5b670459ef2684b318aee/transport/control.go#L336)):*
> ```Go
> }
> 
> func (qb *quotaPool) GetOutFlowWindow() int64 {
> ```

Does this need to be exported?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/lyuxuan/grpc-go/1)*
<!-- Sent from Reviewable.io -->

lyuxuan committed:
fix minor bug

lyuxuan committed:
fix minor bug

MakMukhi committed:
Fix flaky test: TestCloseConnectionWhenServerPrefaceNotReceived (#1870)

* Atomically update minConnectTimeout in test.

* Refactor the flaky test.

* Post review update

* mend

lyuxuan committed:
refator channelz internal impl

lyuxuan committed:
refator channelz internal impl

lyuxuan committed:
Merge branch 'channelz2' into channelz4

lyuxuan committed:
add dummyEntry. fix mintor bug

lyuxuan committed:
improve comments

cesarghali committed:
ALTS: Simplify "New" APIs (#1895)

ALTS NewClientALTS and NewServerALTS APIs are in the grpc/credentials/alts package. Having the term ALTS in the API names is redundant. This PR removes it.

cesarghali committed:
interop: Add use_alts flag to client and server binaries (#1896)

MakMukhi committed:
Fix test race: Atomically access minConnecTimout in testing environment. (#1897)

lyuxuan committed:
only remove server once

lyuxuan committed:
unexport flowcontrol getter func

jhump committed:
credentials/alts: fix infinite recursion bug [in custom error type] (#1906)

lyuxuan committed:
Merge branch 'channelz2' into channelz4

lyuxuan committed:
enable benchmarking channelz

cesarghali committed:
credentials/alts: change ALTS protos to match the golden version (#1908)

jeanbza committed:
server: Convert all non-status errors to codes.Unknown (#1881)

- convertCode utilized errors that were not allowed by the library per https://github.com/grpc/grpc/blob/9d0bc30edbe14fef58f32e74009dd513dee2cfd0/doc/statuscodes.md
- Relevant issue: #1672

menghanl committed:
resolver: always fall back to default resolver when target does not follow URI scheme (#1889)

Previously, any target with "://" would be handled according to the URI scheme even though it did not contain a third slash.

lyuxuan committed:
rebase with channelz funcs.go and type.go

lyuxuan committed:
Merge branch 'master' into channelz4

lyuxuan commented:
Review status: 3 of 31 files reviewed at latest revision, 4 unresolved discussions, some commit checks failed.

---

*[clientconn.go, line 872 at r1](https://reviewable.io:443/reviews/lyuxuan/grpc-go/1#-L6U1sLG0yfTWEBMTokj:-L6UGYcykw5y_Jq-ySJt:b-58b9t6) ([raw file](https://github.com/lyuxuan/grpc-go/blob/6a52ab13a3232d9f8ed5b670459ef2684b318aee/clientconn.go#L872)):*
<details><summary><i>Previously, menghanl (Menghan Li) wrote…</i></summary><blockquote>

Why is this function needed? It looks a bit wired to me even for tests...

And this shouldn't be part of the channelz public interface.
</blockquote></details>

There is race condition for reading and writing channelzID.
But I've fixed the race inside the channelz internal refactoring.

---

*[picker_wrapper.go, line 67 at r1](https://reviewable.io:443/reviews/lyuxuan/grpc-go/1#-L6U3y9QOM-njGIkXGZ-:-L6UO7sFvFkrmIvKHQIS:bzgtbbe) ([raw file](https://github.com/lyuxuan/grpc-go/blob/6a52ab13a3232d9f8ed5b670459ef2684b318aee/picker_wrapper.go#L67)):*
<details><summary><i>Previously, menghanl (Menghan Li) wrote…</i></summary><blockquote>

The `io.EOF` check should not be necessary any more.
</blockquote></details>

Can you explain why? 

---

*[stream.go, line 119 at r1](https://reviewable.io:443/reviews/lyuxuan/grpc-go/1#-L6U6uzCXunkmvNMM80V:-L71Ms6QbbL5jm_I7Ahj:bb16ccj) ([raw file](https://github.com/lyuxuan/grpc-go/blob/6a52ab13a3232d9f8ed5b670459ef2684b318aee/stream.go#L119)):*
<details><summary><i>Previously, menghanl (Menghan Li) wrote…</i></summary><blockquote>

Is it possible to move cc.incrStart/cc.incrEnd into ac.incrStart/ac.incrEnd?
</blockquote></details>

No. retry may make addrconn has more calls started than clientconn

---

*[transport/control.go, line 336 at r1](https://reviewable.io:443/reviews/lyuxuan/grpc-go/1#-L6U83kI6gEy7XgGYt67:-L71MpvWoYit1xjIpJXb:b8tjmas) ([raw file](https://github.com/lyuxuan/grpc-go/blob/6a52ab13a3232d9f8ed5b670459ef2684b318aee/transport/control.go#L336)):*
<details><summary><i>Previously, menghanl (Menghan Li) wrote…</i></summary><blockquote>

Does this need to be exported?
</blockquote></details>

no. unexported

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/lyuxuan/grpc-go/1)*
<!-- Sent from Reviewable.io -->
